### PR TITLE
WIP: Add stream method to channel

### PR DESCRIPTION
**Pull request by jean-airoldie**
This method returns a type which implements Future::Stream.

### CODE REVIEW

**Comment by jean-airoldie** [comment]
Hum, seems like I ran rustfmt which explains most of the modificiations. Is there a specific reason why the repo is not rustfmt'ed?

File: src/channel/mpmc.rs [hunk 1]
```diff
@@ -363,51 +379,120 @@ where A: RingBuf<Item=T>
...
+        let fut = mut_self.future.as_mut().unwrap();
+
+        // Safety: Here we guarantee that the pinned future will not move until
+        // it resolves.
+        let poll = unsafe {
+            let pin_fut = Pin::new_unchecked(fut);
+            pin_fut.poll(cx)
+        };
+
+        // Future has not yet resolved, keep it.
```
Comment (Matthias247) [inline src/channel/mpmc.rs]: Maybe add: "Otherwise drop it"

Otherwise the comment is exactly the opposite of what happens below

File: src/channel/mpmc.rs [hunk 2]
```diff
@@ -363,51 +379,120 @@ where A: RingBuf<Item=T>
...
     }
 }
 
+/// A stream that receives from a `GenericChannel`.
+#[derive(Debug)]
+pub struct ChannelStream<'a, MutexType: RawMutex, T, A>
+where
+    A: RingBuf<Item = T>,
+{
+    channel: Option<&'a GenericChannel<MutexType, T, A>>,
```
Comment (Matthias247) [inline src/channel/mpmc.rs]: Why is that an `Option`?

File: src/channel/mpmc.rs [hunk 3]
```diff
@@ -363,51 +379,120 @@ where A: RingBuf<Item=T>
...
+    future: Option<ChannelReceiveFuture<'a, MutexType, T>>,
```
Comment (Matthias247) [inline src/channel/mpmc.rs]: That definitely seems the way to go for a `Future` -> `Stream` adapter. From a pinning point of view it looks correct, since the `Stream` also needs to be pinned.

What I'm however not yet sure is whether it's ok from a logical point of view. If a user would start polling on the `Stream` but not getting an item, the `Future` would still be stuck in a signaled state in the Stream. If that prevents other waiters on the `Stream` from getting woken up it might be problematic. I think I would need to review the acutal wakeup logic to understand the impact - e.g. whether the `Channel` kept track of `Future`s it notified but which did not yet pick up an item. The `Mutex` implementation definitely does this in order to guarantee fairness. The  `Channel` might need for rendevousz semantics? 

It had been a while since I last looked into it. Will do on the weekend.

File: src/channel/mpmc.rs [hunk 4]
```diff
@@ -363,51 +379,120 @@ where A: RingBuf<Item=T>
...
+    }
+}
+
+impl<'a, MutexType, T, A> FusedStream for ChannelStream<'a, MutexType, T, A>
+where
+    A: RingBuf<Item = T>,
+    MutexType: RawMutex,
+{
+    fn is_terminated(&self) -> bool {
+        self.channel.unwrap().inner.lock().is_closed()
```
Comment (Matthias247) [inline src/channel/mpmc.rs]: That's unfortuntaely a lot more costly then the `is_terminated`  check on the `Future`s -> since it is also doing something different. Do other `FusedStream` implementations also check whether the Stream is fully closed? Should it check whether only this "view" on the stream is closed? That could be achieved by setting `channel = None` when the first receive returns `None` and then always returning the cached value.

File: tests/mpmc_channel.rs [hunk 5]
```diff
@@ -368,6 +387,38 @@ macro_rules! gen_mpmc_tests {
...
+                assert_eq!(count, 1);
+
+                let fut = stream.next();
+                pin_mut!(fut);
+                assert!(fut.as_mut().poll(cx).is_pending());
+                assert_eq!(count, 1);
+
+                channel.close();
+
+                assert_receive_stream_done(cx, &mut fut, None);
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: Maybe add another one of these. I think any following receive call should also return `None`

File: tests/mpmc_channel.rs [hunk 6]
```diff
@@ -138,14 +137,34 @@ macro_rules! gen_mpmc_tests {
                 assert!(receive_fut.as_mut().is_terminated());
             }
 
+            // A stream future shouldn't terminate until the stream
+            // terminates.
+            fn assert_receive_stream_done<FutureType, T>(
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: Since that method is 99.9% of the one above, it seems like one should call the other, and then just add the `is_terminated()` line. I'm however not yet decided on a good name for both.

However if we rather test the `Stream` directly then via `StreamExt` then this most likely needs to stay anyway, and some things would need to get renamed (`FutureType` -> `StreamType`).

File: tests/mpmc_channel.rs [hunk 7]
```diff
@@ -368,6 +387,38 @@ macro_rules! gen_mpmc_tests {
...
+            #[test]
+            fn stream_smoke_test() {
+                let channel = ChannelType::new();
+                let (waker, count) = new_count_waker();
+                let cx = &mut Context::from_waker(&waker);
+
+                let stream = channel.stream();
+                pin_mut!(stream);
+
+                let fut = stream.next();
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: This tests through the `StreamExt` trait, right? Maybe it would be better to directly check the lower level `Stream` APIs in order to exclude the behavior from the extension functions from the test.

File: tests/mpmc_channel.rs [hunk 8]
```diff
@@ -661,28 +712,30 @@ macro_rules! gen_mpmc_tests {
...
 #[cfg(feature = "std")]
 mod if_std {
     use super::*;
-    use futures_intrusive::channel::{Channel, UnbufferedChannel, shared::channel,
-        shared::Sender, shared::Receiver,
-        shared::ChannelReceiveFuture, shared::ChannelSendFuture};
+    use futures_intrusive::channel::{
+        shared::channel, shared::ChannelReceiveFuture, shared::ChannelSendFuture, shared::Receiver,
+        shared::Sender, Channel, UnbufferedChannel,
+    };
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: There is no test for the `Stream` retrieved from the `shared::Receiver`, right? Would be nice to have one, in order to avoid that this one gets accidentally broken.

File: src/channel/mpmc.rs (hunk 1 above)
Comment (jean-airoldie) [reply src/channel/mpmc.rs]: Typo : `if Future has not yet resolved, keep it.`

File: src/channel/mpmc.rs (hunk 4 above)
Comment (jean-airoldie) [reply src/channel/mpmc.rs]: I think the ideal would be a combination of the two. Check if the channel is `None`, if not then check the locked channel.

**Comment by jean-airoldie** [comment]
I'll write a separate PR for `rustfmt` and i'll add a travis.yml which will automatically run it. Then you can [tweak the linter](https://rust-lang.github.io/rustfmt/) to you liking.

**Comment by jean-airoldie** [comment]
If you merge said `rustfmt` PR i'll rebase against it so that the current PR becomes more readable.

File: src/channel/mpmc.rs (hunk 4 above)
Comment (Matthias247) [reply src/channel/mpmc.rs]: I think the first is sufficient. `FusedStream` is mainly used for `select!` in order to prevent it from running in an endless loop in case you use it in loops. Therefore it needs to make sure to avoid streams and futures which have already finished in the following iterations. If you however are not 100% sure and would do one more iteration, it wouldn't be too bad. You would just having wasted one poll cycle.  The same would happen anyway if the stream would get successfully across the `is_terminated()` check and then would get closed before `select!` actually would poll it.

The good part about only checking the `Option` is that we would not need to pay 2x the price of locking the mutex on each `select` operation. Most operations on a stream would be working against a still opened stream - therefore we should optimize for this scenario.

Comment (jean-airoldie) [reply src/channel/mpmc.rs]: Right, makes sense.

Btw, do you know how expensive async mutex calls are (yours for instance)? I'm assuming they are as expensive as OS mutexes but I never measured.

File: tests/mpmc_channel.rs (hunk 7 above)
Comment (jean-airoldie) [reply tests/mpmc_channel.rs]: Agreed, I'm just not familiar with it.

File: src/channel/mpmc.rs (hunk 2 above)
Comment (jean-airoldie) [reply src/channel/mpmc.rs]: Now used by `is_terminated` call.

File: tests/mpmc_channel.rs (hunk 5 above)
Comment (jean-airoldie) [reply tests/mpmc_channel.rs]: Done. I should probably add more meaningful tests however.

File: src/channel/mpmc.rs (hunk 4 above)
Comment (Matthias247) [reply src/channel/mpmc.rs]: Is the question about this particular mutex? Because it's a synchronous one, which just protects the internal state of the channel and is not held for a long time.

The async Mutex in the library has double the cost of a synchronous one in the good case: Noone else has taken the lock before. Then we need 4 atomic operations instead of 2. In the contention case things are a bit more complicated. There is a good part there, and a bad part: The good part is we likely still can get the synchronous Mutex, and don't have to wait on an OS function to get scheduled again. The bad part is now just leave the Waker registered, but go all the way back to the scheduler, which retries the task at some point of time. So it's likely also quite a bit less efficient - but will depend on the executor.

Comment (jean-airoldie) [reply src/channel/mpmc.rs]: Cool. I was wondering if it still did a syscall in the worst case.

File: src/channel/mpmc.rs (hunk 3 above)
Comment (jean-airoldie) [reply src/channel/mpmc.rs]: So a scenario would be user create a stream, polls the stream once, which returns `Poll::NotReady` then drops the stream future. The inner future is stuck in the stream and might be blocking subsequent futures.

Comment (Matthias247) [reply src/channel/mpmc.rs]: Right. This can also happen with the `Future` based API. But there you are a lot more likely to see that an old Future is still around - especially since you can't move the Futures. In the Streams it is somewhat hidden. 

But now since I wrote that: At least the Stream also needs to be pinned and can not be moved - so the user moving around the Stream without knowing that some registration is still active is not possible.

File: src/channel/mpmc.rs (hunk 4 above)
Comment (Matthias247) [reply src/channel/mpmc.rs]: In the worst worst case you would. There you would try to lock the synchronous Mutex exactly in the moment when another thread has it locked. But that probability is low. The async Mutex works like this:

1. Take sync lock
2. Check if some other task has marked themself as owner.
  No -> Mark as owner. Goto 5
  Yes -> Insert the wait node for the current task in the waiter list. Goto 4
3. Unlock sync lock

4. Yield the current task, and wait until the executor schedules it again (by another task calling `.wake()`. Then go back to 1.

5. Use data
6. Take sync lock
7. Mark as no owner. potentially set someone else as owner, and/or wake up another task
8. Unlock sync lock

**Comment by Matthias247** [comment]
I have reviewed the code again. As mentioned, there could be an issue when someone starts reading from the `Stream` but then doesn't complete the read (doesn't extract a value). When the read is started the `Future` is registered in the Channel in the waiters list. When someone wants to send a value, the laster waiter in the waiter list is signalled [here](https://github.com/Matthias247/futures-intrusive/blob/a155cbefc556f45818dcbee6892b6b7f623cdfef/src/channel/mpmc.rs#L135). This could be e.g. the `Future` which is stored in the `Stream`. Other tasks are not signalled.

That means e.g.
- if the channel capacity is 1
- 2 tasks are waiting on the channel
- the first (`Stream`) task gets signalled, but doesn't read from the channel anymore

then the second task will wait for the notification until the `Stream` gets dropped (which leads to waking up another waiter). 

If during the situation another producer would send something to the channel, it would
- insert that sender into the wait list
- wakeup the old second task

Then that task would actually extract the old value, and the value from the second sender is shifted into the channel buffer.

Therefore it seems like we might only lose the buffer and not run into a permanent deadlock state, but I'm still not 100% sure. Need to think further about other edge-cases.

In order to avoid the issue totally each send could wake-up all stored receivers instead of only a single one. However that would introduce a lot of churn on normally used channels with multiple producers and consumers. E.g. if there are 5 consumers
- they all would register at the channel and wait
- a single send would unblock all of them
- only consumer gets the value
- at least 4 consumers register themself again at the channel

which means lots of churn in the executor. Therefore I'm not too keen on implementing a change to wake-up all receivers.

**Comment by jean-airoldie** [comment]
Lets say we did would run into a permanent deadlock, couldn't we circumvent it with a drop implementation? For instance, by retrieving the value on drop by polling one last time?

**Comment by Matthias247** [comment]
Where would you `drop`? The `ChannelReceive/SendFuture` which you store in the `Stream` already implements `Drop` and wakes up another waiter if it had been notified, but the notification not been used. So if you drop the stream everything is good again.

The concern here is mainly that there is some hidden state in the `Stream` which might cause some none obvious behavior for some users. When someone uses the pure `Future` based APIs it is a bit more obvious, since the `Future` is still visible / in scope. When that gets dropped also everything is good again.

**Comment by jean-airoldie** [comment]
I did some very basic testing in https://github.com/Matthias247/futures-intrusive/pull/14/commits/e96b7ff044ea7769bce166aa94de39da377e2f9b and it seems like for the low level future API its not a big deal. From what I understand a forgotten future (send & receive) will eat up one wakeup, but everything works as normal after that. The values stored in the buffer won't be affected so its basically like a wakeup was lost in the aether.

**Comment by Nemo157** [comment]
> so its basically like a wakeup was lost in the aether.

Lost wakeups can lead to deadlocks if the task is not concurrently doing anything else or never encounters a spurious wakeup. That could be allowable though, leaking a future/stream instead of dropping it already sort of violates the expected usage.

I'll try and take a look at this in a bit.

**Comment by Matthias247** [comment]
I think I now have better understanding on what will happen regarding the lost wakeups: Every lost waker will cause one sender to be blocked. However on each consecutive `send()` operation another receiver will be woken up. Independent of whether the sent element went into the internal buffer or the sender got attached to the waiter list. That will unblock a waiting receiver. 

That means in my understanding the maximum amount each non-used wakeup would cause an exactly on element in the channel being not used, but successive reads and writes should be succesful. That means a latency of 1 element would be introduced into the system. I think that might be acceptable in typical MPMC workflows - where elements are unrelated and latency is likely unpredictable. In single consumer workflows this should never happen, since that would mean the only consumer would not read from the channel. For 

To answer the question from yesterday why we don't run into the debug assert: The receivers always read from elements in the buffer. While reading one element from the buffer, they then also copy one element from the oldest send waiter into the buffer in order to unblock one more sender => There is no direct handover from a sender to the receiver. The buffer is always used in between.

**Comment by jean-airoldie** [comment]
>  That means a latency of 1 element would be introduced into the system. I think that might be acceptable in typical MPMC workflows

For a continuous workflow this is acceptable indeed. For temporary workflow, one of the sender would close the channel after its done which would wakes up all waiters.

File: src/channel/mpmc.rs [hunk 9]
```diff
@@ -500,6 +524,80 @@ where
...
+        let mut_self: &mut Self = unsafe { Pin::get_unchecked_mut(self) };
+        match mut_self.channel.take() {
+            Some(channel) => {
+                // Poll the next element.
+                if mut_self.future.is_none() {
+                    mut_self.future.replace(channel.receive());
+                }
+                let fut = mut_self.future.as_mut().unwrap();
+
+                // Safety: Here we guarantee that the pinned future will not move until
```
Comment (Matthias247) [inline src/channel/mpmc.rs]:
```suggestion
                // Safety: We guarantee that the pinned future will not move until
```

File: src/channel/mpmc.rs [hunk 10]
```diff
@@ -500,6 +524,80 @@ where
...
+                // it resolves.
```
Comment (Matthias247) [inline src/channel/mpmc.rs]:
```suggestion
                // it resolves by storing it as part of the pinned `Stream`
```

File: src/channel/mpmc.rs (hunk 9 above)
Comment (Matthias247) [reply src/channel/mpmc.rs]: Just a nit, because we don't apply the guarantee in this line (as the comment indicates), but just rely on this properties in order to be able to call `Pin::new_unchecked`.

File: src/channel/mpmc.rs [hunk 11]
```diff
@@ -833,6 +931,95 @@ mod if_alloc {
...
+                    unsafe { Pin::get_unchecked_mut(self) };
+                match mut_self.receiver.take() {
+                    Some(receiver) => {
+                        // Poll the next element.
+                        if mut_self.future.is_none() {
+                            mut_self.future.replace(receiver.receive());
+                        }
+                        let fut = mut_self.future.as_mut().unwrap();
+
+                        // Safety: Here we guarantee that the pinned future will not move until
```
Comment (Matthias247) [inline src/channel/mpmc.rs]:
```suggestion
                        // Safety: We guarantee that the pinned future will not move until
```

File: src/channel/mpmc.rs [hunk 12]
```diff
@@ -833,6 +931,95 @@ mod if_alloc {
...
+                        // it resolves.
```
Comment (Matthias247) [inline src/channel/mpmc.rs]:
```suggestion
                        // it resolves by storing it as part of the pinned `Stream`
```

**Comment by jean-airoldie** [comment]
I was away for a while and I forgot about this PR.

What do you think would be necessary before this can be merged? What kind of additional tests do you think would be needed?

**Comment by Matthias247** [comment]
Sorry for the late feedback. I convinced myself that this is good to have that way. If we just degenerate on buffers instead of locking it seems acceptable. And @Nemo157's comment that even locking might be acceptable due to the fact that not using a `Stream` might be treated as misbehavior is interesting.  Although I prefer to let people not run into surprises - which then lead to "why doesn't that work" tickets posted :-)

One random thought I had was whether the `Stream` support could be behind a feature flag. That way we could increase the chance that people read something about the feature in a readme instead of just trying to use APIs. But on the other hand just having a `feature` for this API seems weird. Maybe I would rather have one which disables all dependencies on `future-core`.

Another random thought I had is that the `Stream` support could also be implemented in another fashion: Instead of storing a `Future` in the `Stream` it could also directly store a `ListNode`. That would make it probably a tiny bit more compact. But on the other hand delegating it to the `Future` internally seems like less code to maintain, so I guess I have a preference for the path taken.

File: tests/mpmc_channel.rs [hunk 13]
```diff
@@ -918,6 +941,41 @@ macro_rules! gen_mpmc_tests {
...
+
+                let stream = channel.stream();
+                pin_mut!(stream);
+
+                assert!(stream.as_mut().poll_next(cx).is_pending());
+                assert_eq!(count, 0);
+
+                assert_send(cx, &channel, 1);
+
+                assert_next_done(cx, &mut stream, Some(1));
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: You could try reading multiple values here. I guess just a loop which covers line 954-960 for some iterations might be fine.

File: tests/mpmc_channel.rs [hunk 14]
```diff
@@ -918,6 +941,41 @@ macro_rules! gen_mpmc_tests {
...
+            #[test]
+            fn stream_smoke_test() {
+                let channel = ChannelType::new();
+                let (waker, count) = new_count_waker();
+                let cx = &mut Context::from_waker(&waker);
+
+                let stream = channel.stream();
+                pin_mut!(stream);
+
+                assert!(stream.as_mut().poll_next(cx).is_pending());
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: Maybe let's add another test which checks for the "immediately ready" variant (send before receive).

File: tests/mpmc_channel.rs [hunk 15]
```diff
@@ -918,6 +941,41 @@ macro_rules! gen_mpmc_tests {
                 assert_send_done(cx, &mut send_fut2, Ok(()));
                 assert_send_done(cx, &mut send_fut3, Ok(()));
             }
+
+            #[test]
+            fn stream_smoke_test() {
+                let channel = ChannelType::new();
```
Comment (Matthias247) [inline tests/mpmc_channel.rs]: This just checks the non-shared variant. We should also have a test for the shared version.

**Comment by jean-airoldie** [comment]
It would be interesting it we could add a compiler warning if the user misuses the stream. Kinda like `#[must_use]` but instead generates a warning when a stream is not driven to completion.

**Comment by Matthias247** [comment]
> I was away for a while and I forgot about this PR.
> 
> What do you think would be necessary before this can be merged? What kind of additional tests do you think would be needed?

Ah, good point around testing. I just reviewed again what this has and have added a few suggestions based on it. Apart from that I think just checking what the other channel tests have, and whether we miss an importatant edge condition would be good. But since we just delegate to the existing stream most things should be pretty well covered.

**Comment by Matthias247** [comment]
> 
> 
> It would be interesting it we could add a compiler warning if the user misuses the stream. Kinda like `#[must_use]` but instead generates a warning when a stream is not driven to completion.

That sounds cool. But I have no idea how, since  it's a dynamic property. It would somehow need to track that an operation returned `Pending` and the object wasn't used from then on until `Drop`.  And event that is a legitimate use-cases inside `select!` : The operation could just have been cancelled because it wasn't required to run to completion. Therefore time plays a major role in the decision.

I guess runtime checks could be used: When the object last retuns `Poll` store a timestamp. On `drop()` compare the timestamps and if a "long" timeframe has exceeded emit a warning (or `debug_assert!`). But no idea for compile-time check possibilities.

**Comment by jean-airoldie** [comment]
I also added some very vague doc concerning the `SharedStream` & `ChannelStream` to warn the user against potential lost wake-up notifications.

**Comment by Matthias247** [comment]
I merged this now. Thanks for working on it!
